### DISCUSSION TITLE
Howto mkfs.ext4/create filesystem into image file?

### DISCUSSION THREAD

**Discussion by andig** [community/community#27156] [Actions]
<p>Build task needs to create an OS image. Image is written to file, now need to add ext4 partition (or vfat or anything really) to the image.</p>
<p>Is there a way to do that? Mounting the image using loopback fails.</p>

**Comment by airtower-luna** [comment]
<p>You need to create the filesystem before you can mount it. You can apply <code>mkfs.ext4</code> directly to the file if you only need a filesystem image.</p>
<p>If you need to set up partitions within the image file you need to set up the loop device first using <code>losetup</code> and partition it. Note that to actually use partitions on a loop device the <code>loop</code> kernel module must be mounted with a suitably large <code>max_part</code> parameter.</p>

**Comment by andig** [comment]
<p>Thank you for the quick answer. I already have the image file containing 3 partitions (boot, root, unallocated). I understand I need to <code>mkfs.ext4</code> on the unallocated partition once the image file is mounted.</p>
<p>I’m still failing to do so on CI- any hint would be appreciated.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>You can use command like this to set up the loop device:</p>
<pre><code class="lang-bash">sudo losetup --partscan /dev/loop0 disk.img
</code></pre>
<p>The <code>--partscan</code> parameters enables detection of partitions in the loop device, <code>/dev/loop0</code> is the loop device you want to use. You might need to use another if you already use other loop devices, or use <code>-f</code> to automatically use a free one (but then you need to find which one).</p>
<p>The example command should give you additional devices for the partitions, named like <code>/dev/loop0p1</code> (with 1 being the partition number here). You can work with those like with any disk partition, e.g. run <code>mkfs.ext4 /dev/loop0p3</code>.</p>
